java/ql/lib/semmle/code/java/regex/regex.qll

```diff
@@ -0,0 +1,933 @@
+
+  /** Holds if a quoted sequence is found between `start` and `end` */
+  predicate quote(int start, int end) { this.quote(start, end, _, _) }
+
+  /** Holds if a quoted sequence is fund between `start` and `end`, with ontent found between `inner_start` and `inner_end`. */
+  predicate quote(int start, int end, int inner_start, int inner_end) {
+    exists(int index |
+      this.quoteDelimiter(index, start) = true and
+      (
+        index = 1
+        or
+        this.quoteDelimiter(index - 1, _) = false
+      ) and
+      inner_start = start + 2 and
+      inner_end = end - 2 and
+      inner_end > inner_start and
+      this.quoteDelimiter(inner_end) = false and
+      not exists(int mid |
+        this.quoteDelimiter(mid) = false and mid in [inner_start .. inner_end - 1]
+      )
+    )
+  }
+
+  /**
+   * A control sequence, `\cx`
+   * `x` may be any ascii character including special characters.
+   */
+  predicate controlEscape(int start, int end) {
+    this.escapingChar(start) and
+    this.getChar(start + 1) = "c" and
+    end = start + 3
+  }
+
+  /** Gets the text of this regex */
+  string getText() { result = this.(StringLiteral).getValue() }
+
+  string getChar(int i) { result = this.getText().charAt(i) }
+
+  string nonEscapedCharAt(int i) {
+    result = this.getText().charAt(i) and
+    not exists(int x, int y | this.escapedCharacter(x, y) and i in [x .. y - 1]) and
+    not exists(int x, int y | this.quote(x, y) and i in [x .. y - 1]) and
+    not exists(int x, int y | this.controlEscape(x, y) and i in [x .. y - 1])
+  }
+
+  private predicate isOptionDivider(int i) { this.nonEscapedCharAt(i) = "|" }
+
+  private predicate isGroupEnd(int i) { this.nonEscapedCharAt(i) = ")" and not this.inCharSet(i) }
+
+  private predicate isGroupStart(int i) { this.nonEscapedCharAt(i) = "(" and not this.inCharSet(i) }
+
+  predicate failedToParse(int i) {
+    exists(this.getChar(i)) and
+    not exists(int start, int end |
+      this.topLevel(start, end) and
+      start <= i and
+      end > i
+    )
+  }
+
+  /** An escape sequence that includes braces, such as named characters (\N{degree sign}), named classes (\p{Lower}), or hex values (\x{h..h}) */
+  private predicate escapedBraces(int start, int end) {
+    this.escapingChar(start) and
+    this.getChar(start + 1) = ["N", "p", "P", "x"] and
+    this.getChar(start + 2) = "{" and
+    this.getChar(end - 1) = "}" and
+    end > start and
+    not exists(int i | start + 2 < i and i < end - 1 | this.getChar(i) = "}")
+  }
+
+  /**
+   * Holds if an escaped character is found between `start` and `end`.
+   * Escaped characters include hex values, octal values and named escapes,
+   * but excludes backreferences.
+   */
+  predicate escapedCharacter(int start, int end) {
+    this.escapingChar(start) and
+    not this.backreference(start, _) and
+    (
+      // hex value \xhh
+      this.getChar(start + 1) = "x" and
+      this.getChar(start + 2) != "{" and
+      end = start + 4
```

Comment:
Should the individual hex chars be checked?

Reply:
I'm not sure that would be worth it - doing so would only help to reject invalid regexes (which we don't expect there to be in real databases); and wouldn't help accept more valid regexes.